lib/std/heap.zig

```diff
@@ -21,67 +21,119 @@ pub const GeneralPurposeAllocator = @import("heap/general_purpose_allocator.zig"
 
 const Allocator = mem.Allocator;
 
-usingnamespace if (comptime @hasDecl(c, "malloc_size"))
-    struct {
-        pub const supports_malloc_size = true;
-        pub const malloc_size = c.malloc_size;
-    }
-else if (comptime @hasDecl(c, "malloc_usable_size"))
-    struct {
-        pub const supports_malloc_size = true;
-        pub const malloc_size = c.malloc_usable_size;
+const CAllocator = struct {
+    comptime {
+        if (!builtin.link_libc) {
+            @compileError("C allocator is only available when linking against libc");
+        }
     }
-else
-    struct {
-        pub const supports_malloc_size = false;
-    };
 
-pub const c_allocator = &c_allocator_state;
-var c_allocator_state = Allocator{
-    .allocFn = cAlloc,
-    .resizeFn = cResize,
-};
+    usingnamespace if (comptime @hasDecl(c, "malloc_size"))
+        struct {
+            pub const supports_malloc_size = true;
+            pub const malloc_size = c.malloc_size;
+        }
+    else if (comptime @hasDecl(c, "malloc_usable_size"))
+        struct {
+            pub const supports_malloc_size = true;
+            pub const malloc_size = c.malloc_usable_size;
+        }
+    else
+        struct {
+            pub const supports_malloc_size = false;
+        };
 
-fn cAlloc(self: *Allocator, len: usize, ptr_align: u29, len_align: u29, ret_addr: usize) Allocator.Error![]u8 {
-    assert(ptr_align <= @alignOf(c_longdouble));
-    const ptr = @ptrCast([*]u8, c.malloc(len) orelse return error.OutOfMemory);
-    if (len_align == 0) {
-        return ptr[0..len];
-    }
-    const full_len = init: {
-        if (supports_malloc_size) {
-            const s = malloc_size(ptr);
-            assert(s >= len);
-            break :init s;
+    // The alignment guaranteed by malloc, the value matches the result of the C
+    // expression `alignof(max_alignment_t)`
+    const min_ptr_alignment = comptime std.math.max(
```

Comment:
Is this guaranteed for small allocations? I have an inkling that e.g. allocating a single byte with malloc is allowed to have an alignment of `1`.

Reply:
The [standard](https://port70.net/~nsz/c/c11/n1570.html) is not that clear on that front, that point was recently [amended]( www.open-std.org/jtc1/sc22/wg14/www/docs/n2293.htm) not to be misunderstood by the reader (See [C2X](https://port70.net/~nsz/c/c2x/n2434.pdf) 7.22.3).

The check for using malloc could be amended as `alignment <= min_ptr_alignment and alignment <= len` to be safe with "strong-alignment" and "weak-alignment" mallocs.

Reply:
Thanks for the great references! From that N2293 I think we should only assume "weak-alignment"; aside from possibly when statically linking against a libc known to have strong-alignment (the only circumstance of which seems to be musl).

Reply:
I think there may be a problem in `realloc`. On windows you can't mix&match regular and aligned malloc/free (the aligned alloc/free is just a [thin wrapper over regular malloc](https://github.com/Alexpux/mingw-w64/blob/master/mingw-w64-crt/misc/mingw-aligned-malloc.c) and so the same choice made in `alloc` must be done in `realloc`.
The problem is with the `len` parameter, the `len_align` factor may change the slice size and it'd be impossible for us to pick the correct free function. On the other systems using `posix_memalign` you can use `free` and call it a day.

Reply:
Would an answer be to always prefer `_aligned_` versions on windows when possible ambiguity could occur?

Reply:
Using the aligned variant incurs in a IMO negligible overhead, it may be a solution.